### ISSUE TITLE
Add a warning when `prefetch_factor * num_worker <= batch_size`

### ISSUE THREAD

**Issue by robin5540** [pytorch/pytorch#58030]
## 🚀 Feature

Show a warning in `torch.util.data.dataloader.Dataloader` when `prefetch_factor * num_worker <= batch_size`

## Motivation

The goal of using mp in dataloader is to avoid waiting for data. When the fetch buffer is smaller than `batch_size`, `next(dl_iter)` will end up always waiting for data. This is quite an annoying bug when misconfigured.

## Pitch

Adding a warning will make this obscure case easier to debug.

## Alternatives

Throw an error: probably isn't the best idea

Automatically force `prefetch_factor = max(batch_size // num_worker, prefetch_factor)`: this might yield confusing OOM.

## Additional context

Will send a PR

cc @SsnL @VitalyFedyunin @ejguan

**Comment by ejguan** [comment]
> When the fetch buffer is smaller than `batch_size`, `next(dl_iter)` will end up always waiting for data

Is that true? IIRC, each worker should have a buffer with size of `prefetch_factor * batch_size`.

**Comment by robin5540** [comment]
@ejguan

When priming the loop: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L973-L975
When put new work on a worker: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L1206

I think going to `prefetch_factor * batch_size` may make memory size hard to control. If going to `prefetch_factor * batch_size`, I think it might be a good idea to support fractional `prefetch_factor`

**Comment by ejguan** [comment]
@robin5540 
Here is the inner logic to put indexes, where a batch of indexes are sampled and sent to worker if `batch_size` is set.
https://github.com/pytorch/pytorch/blob/d212bf186390fdcc3b36cd3b41489544c4f50b75/torch/utils/data/dataloader.py#L511-L512
https://github.com/pytorch/pytorch/blob/94bb1150a73e005aa23aaf13fecad79d18fa50e3/torch/utils/data/dataloader.py#L366-L375
Like I said, `prefetch_factor` is used for **every single worker** to have `prefetch_factor * batch_size` buffer. So fractional `prefetch_factor` doesn't make sense, right?

**Comment by robin5540** [comment]
I see. `prefetch_factor * batch_size` is only true when using a batch sampler. `len(buffer) < len(batch_size)` can still be true when use `sampler`, `shuffle`, and manual `batch_size`

**Comment by robin5540** [comment]
@ejguan 
Looks like the warning should be `not batch_sample and prefetch_factor * num_worker <= batch_size`. wdyt?

**Comment by ejguan** [comment]
I am sorry that I am not sure about manual `batch_size` you are referring.

> Looks like the warning should be `not batch_sampler and prefetch_factor * num_worker <= batch_size`

Each worker will fetch a whole `batch_size` of data for each time. I don't think `prefetch_factor` is related to `num_worker`. Each worker has `prefetch_factor * batch_size`, right? So, no chance `prefetch_factor * num_worker <= batch_size` will happen.

**Comment by robin5540** [comment]
Ohhh, I see.
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L274-L276
The sample is always wrapped into a batch sample

**Comment by ejguan** [comment]
Closing the issue. Please feel free to re-open it when needed.
And feel free to discuss questions on PyTorch Forum (https://discuss.pytorch.org/)

**Comment by amin-nejad** [comment]
Thank you, I've created a PR
